### PR TITLE
grpc: serialize-to and parse-from zero copy streams

### LINKED ISSUE #3472: Optimize (de)serialization for Google gRPC client
Today, we do multiple copies to (de)serialize protos to byte buffers, see:
* https://github.com/envoyproxy/envoy/blob/396f52de26e8864dbbefa903ff605bd52af11b3e/source/common/grpc/google_async_client_impl.h#L233
* https://github.com/envoyproxy/envoy/blob/396f52de26e8864dbbefa903ff605bd52af11b3e/source/common/grpc/google_async_client_impl.cc#L303

This was due to gRPC library limitations at the time of implementation. @ncteisen has added in support for zero-copy (de)serialization since then, see https://github.com/tensorflow/tensorflow/commit/105d9795ae692ed2486652e5d672825ccbd726e9 for an example of how this is done.

We should cleanup the various TODOs and switch over to this way of doing things.

### PR DESCRIPTION
**Pull request by gumpt**
*Description*:
This change takes advantage of newly-public grpc headers to make fewer copies to/from strings before using them.  Before this change, a message being serial/deserialization would involve copies between protobuf messages, vectors of slices, strings, and grpc byte-buffers.  Now extra step(s) are gone, and data are copied directly between protobuf messages and byte-buffers.

*Risk Level*:
Low due to quality of tests predating this change

*Testing*:
`bazel test //test/common/grpc:grpc_client_integration_test`, `bazel test //test/...`

*Docs Changes*:
N/A

*Release Notes*:
N/A

Fixes #3472

### CODE REVIEW

**Comment by gumpt** [comment]
@htuch please take a look :-)

File: source/common/grpc/google_async_client_impl.h [hunk 1]
```diff
@@ -230,22 +231,18 @@ class GoogleAsyncStreamImpl : public AsyncStream,
...
-    // at the cost of some additional implementation complexity.
-    // Also see
-    // https://github.com/grpc/grpc/blob/5e82dddc056bd488e0ba1ba0057247ab23e442d4/include/grpc%2B%2B/impl/codegen/proto_utils.h#L42
-    // which gives us what we want for zero copy, but relies on grpc::internal details; we can't get
-    // a grpc_byte_buffer from grpc::ByteBuffer to use this.
     PendingMessage(const Protobuf::Message& request, bool end_stream)
-        : slice_(request.SerializeAsString()), buf_(grpc::ByteBuffer(&slice_, 1)),
+        : buf_([](const Protobuf::Message& request) -> absl::optional<grpc::ByteBuffer> {
+            grpc::ByteBuffer buffer;
+            grpc::ProtoBufferWriter writer(&buffer, 1024, request.ByteSize());
```
Comment (lizan) [inline source/common/grpc/google_async_client_impl.h]: Is there any reason `1024` was chosen here?

Comment (gumpt) [reply source/common/grpc/google_async_client_impl.h]: It's only a guess at a sane default for a block size.  I'd prefer to plug in a constant defined somewhere else if you could point me to the right constant and/or place.

Comment (lizan) [reply source/common/grpc/google_async_client_impl.h]: [kProtoBufferWriterMaxBufferLength](https://github.com/grpc/grpc/blob/v1.12.1/include/grpcpp/impl/codegen/proto_buffer_writer.h#L44) is used as default in grpc/grpc:
https://github.com/grpc/grpc/blob/v1.12.1/include/grpcpp/impl/codegen/proto_utils.h#L63

Comment (gumpt) [reply source/common/grpc/google_async_client_impl.h]: Great, thank you!

**Comment by lizan** [comment]
oops I somewhat missed the PR review policy :-/  @htuch let me know if you have any concern on this PR.

**Comment by htuch** [comment]
@lizan the PR LGTM, but yeah, for non-trivial PRs it's ideal to have a senior maintainer take a pass. I know @danielhochman was going to take a look as well. All good in any case..

**Comment by ncteisen** [comment]
Nice change! Was there any performance impact from it?

**Comment by htuch** [comment]
@ncteisen this is control-plane side and we don't do any benchmarking of this as it's not super critical. It might make an impact as you scale to very large config sizes, but I don't think most folks in Envoy land are there yet.

**Pull request merged**